linkerd/stack/src/gate.rs

```diff
@@ -115,24 +123,60 @@ impl Tx {
     }
 
     /// Opens the gate.
-    pub fn open(&self) {
-        if self.0.send(State::Open).is_ok() {
-            debug!("Gate opened");
+    pub fn open(&self) -> Result<(), Closed> {
+        if self.0.is_closed() {
+            return Err(Closed(()));
+        }
+        match self.0.send_replace(State::Open) {
+            State::Open => {}
+            State::Limited(lim) => {
+                debug!("Limited => Open");
+                lim.close();
```

Comment:
I think it might be nice if we documented at the top level that a `Limited` state's semaphore will be closed when the gate transitions back to open --- maybe add a rustdoc comment about that to the `State::Limited` variant?